### ISSUE TITLE
CFP: hubble CLI - from-node-labels + to-node-labels

### ISSUE THREAD

**Issue by saintdle** [cilium/cilium#34133]
Support additional Hubble CLI filters `--from-node-labels` and `--to-node-labels`, now that the filter `--node-labels` is [available](https://github.com/cilium/cilium/pull/32851). 

This would allow the capture of all flows originating or terminating against specific node/s identified by their node labels.

**Comment by kaworu** [comment]
Some Hubble flows fields are directional (i.e. have source and destination) like l3/l4 info, endpoint, etc. and it make sense for their associated filters to have corresponding `--from-field`/`--to-field` filters. However node info (namely `node_name` and `node_labels`) are derived from the node on which the flow is observed without notion of direction.

**Comment by saintdle** [comment]
> Some Hubble flows fields are directional (i.e. have source and destination) like l3/l4 info, endpoint, etc. and it make sense for their associated filters to have corresponding `--from-field`/`--to-field` filters. However node info (namely `node_name` and `node_labels`) are derived from the node on which the flow is observed without notion of direction.

Is there a way to apply some additional logical to the the node label field with other fields to be able to filter as such? For example, if a flow `is_reply=true` then the assumption is that this would be traffic going to a node?

**Comment by kaworu** [comment]
`is_reply` and `traffic_direction` are inferred by Hubble using best-effort heuristics. I think they can be useful for metrics to get a general sense of what is going on, but at the individual flow level they can be misleading (or just wrong).

Even assuming they were always correct, `is_reply` is based on CT and `traffic_direction` is ~based on network interface, neither of which maps well to node. Take for example the flows between two Pods co-located on the same node N: how should the filters behave in that case? I would argue that logically, since all the flows are to and from the same node N, then both `--from-node-labels` and `--to-node-labels` should show all the flows between the two Pods, yet I cannot think of a combination of `is_reply`/`traffic_direction` with `node_labels` that would work.

so tl;dr I don't see how we can implement this CFP flags without adding some src/dst node info to Hubble flows, both of which will require additional lookups (and maybe caching?) on the Hubble side for every flow.

**Comment by chancez** [comment]
We already have to get the remote pod information, getting remote node information wouldn't necessarily be a bad thing. But definitely a bigger task to implement these filters since the requisite information isn't available currently.

**Comment by kaworu** [comment]
@chancez correct, although I'm less worried about the amount of work needed than the overhead. If we could implement the new flags based on already existing functionalities without incurring overhead in the Hubble server (as @saintdle originally suggested) then I'd not be opposed. On the other hand if the patch introduce a feature/overhead trade-off then I think we should first discuss whether we should do it in the first place, starting with some concrete use-cases.

**Comment by saintdle** [comment]
My main use case is to help quickly identify traffic between nodes placed in specific availability zones. This would be to help then use features such as node-local-dns and local redirection policy to help reduce cross-AZ traffic, which does have a nominal cost when using cloud providers such as AWS/Azure/Google.

**Comment by chancez** [comment]
Seems related to https://github.com/cilium/cilium/issues/33601

**Comment by chancez** [comment]
I took a brief look and the `NodeManager` on the Daemon should have access to CiliumNode resources for all nodes in the cluster, which have the node labels from the k8s node. The `failure-domain.beta.kubernetes.io/region` and `failure-domain.beta.kubernetes.io/zone` labels are standardized so everything necessary to get the remote node labels seems to exist from that perspective. However, there needs to be a way to determine which nodes is going to. This would mean getting remote pod (which we already do to provide the destination endpoint information), then getting the node for that pod. So everything should already exist to make this possible, it would just need to be implemented.

**Comment by chancez** [comment]
Looking a little more, it's not terribly hard to implement, but basically all our logic around nodes works on the `LocalNodeStore`, so we would need to do some work to make it more generic to work around arbitrary Cilium nodes. As it happens https://github.com/cilium/cilium/pull/32144 is working towards that goal. I think this could make sense after that's implemented, as it would likely simplify things quite a bit if we had a central store to query and subscribe to all node information updates.

**Comment by DPS0340** [comment]
I'd like to work on this. Reading the thread, the blocker is that flows don't carry source/destination node info, plus @kaworu's constraint that the flags shouldn't add per-flow overhead in the Hubble server.

The building blocks already exist in the agent: the ipcache tracks the host IP owning each endpoint IP (`ipToHostIPCache`), and the node manager already has labels for all cluster nodes (`ParseCiliumNode` copies CiliumNode labels into `node.Node.Labels`). So resolving flow IP → owning node → node labels is two in-memory map lookups — no extra watchers and no k8s API calls.

Two possible shapes:

1. **Enrich flows**: add `source_node_labels`/`destination_node_labels` to the `Flow` proto, populated in the parser via a new `NodeLabelsGetter`. Cost: two map lookups per flow plus larger flows, but it also provides the data #38932 asks for.
2. **Filter-only (lazy)**: add `source_node_labels`/`destination_node_labels` only to `FlowFilter`; the filter func resolves node labels through the getter *only when one of these filters is active*. Zero overhead when the flags aren't used and no `Flow` schema change. Caveats: the filter needs agent-side state so it must be evaluated on the agents (where relay pushes filters anyway), and the resolved labels aren't visible in the flow output.

Option 2 seems to satisfy the no-overhead constraint. If maintainers agree on a direction I'll open a PR — happy to write this up in more detail first if preferred.

**Comment by chancez** [comment]
I don't think it's really worth trying to do this only in the flow filter path because a lot of users have a desire for flows to contain destination node information for other use-cases like cross-AZ cost attribution, so I'd push for the first option.
